File: leetcode/101_SymmetricTree.java
package com.terrykwon;



public class SymmetricTree {
    private static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }

    public boolean isSymmetric(TreeNode root) {
        if (root == null) return true;

        TreeNode l = root.left;
        TreeNode r = root.right;

        return recursiveCheck(l, r);
    }

    private boolean recursiveCheck(TreeNode l, TreeNode r) {
        if (l == null && r == null) {
            return true;
        } else if (l == null || r == null) {
            return false;
        }

        if (l.val != r.val) {
            return false;
        }

        return recursiveCheck(l.left, r.right) && recursiveCheck(l.right, r.left);
    }
}

File: leetcode/101_symmetric_tree.cpp
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        return areSymmetricSubtrees(root->left, root->right);
    }

    bool areSymmetricSubtrees(TreeNode* a, TreeNode* b) {
        if (!a && !b) {
            return true;
        } else if (!a || !b) {
            return false;
        } else if (a->val != b->val) {
            return false;
        } else {
            return areSymmetricSubtrees(a->left, b->right)
                    && areSymmetricSubtrees(a->right, b->left);
        }
    }
};